src/urllib3/response.py

```diff
@@ -785,12 +766,104 @@ def read(
+        return len(self._decoded_bytes)
+
+    def read(
+        self,
+        amt: Optional[int] = None,
+        decode_content: Optional[bool] = None,
+        cache_content: bool = False,
+    ) -> bytes:
+        """
+        Similar to :meth:`http.client.HTTPResponse.read`, but with two additional
+        parameters: ``decode_content`` and ``cache_content``.
 
-            data = self._decode(data, decode_content, flush_decoder)
+        :param amt:
+            How much of the content to read. If specified, caching is skipped
+            because it doesn't make sense to cache partial content as the full
+            response.
+
+        :param decode_content:
+            If True, will attempt to decode the body based on the
+            'content-encoding' header.
 
+        :param cache_content:
+            If True, will save the returned data such that the same result is
+            returned despite of the state of the underlying file object. This
+            is useful if you want the ``.data`` property to continue working
+            after having ``.read()`` the file object. (Overridden if ``amt`` is
+            set.)
+        """
+        if amt is not None:
+            cache_content = False
+
+        if decode_content is None:
+            decode_content = self.decode_content
+
+        if self.length_remaining is None and decode_content:
+            warnings.warn(
+                "The response is missing a Content-Length header. When "
+                "reading with decode_content=True, the HTTPResponse.read "
+                "call may return no bytes or less bytes than specified even "
+                "though there may be still more data to read.",
+                ContentLengthMissing,
+            )
+
+        if not decode_content:
+            data = self._raw_read(amt)
             if cache_content:
                 self._body = data
+            return data
+
+        if amt is not None and self._decoded_bytes_in_buffer() >= amt:
+            return self._popleft_decoded_bytes(amt)
+
+        data = self._raw_read(amt)
+
+        if not data and self._decoded_bytes_in_buffer() == 0:
+            return data
+
+        self._init_decoder()
+
+        if amt is None or (amt != 0 and not data):
+            flush_decoder = True
+        else:
+            flush_decoder = False
+
+        decoded_data = self._decode(data, decode_content, flush_decoder)
+        self._decoded_bytes.extend(decoded_data)
+
+        if (
+            amt is not None
+            and self._decoded_bytes_in_buffer() < amt
+            and self.length_remaining is not None
+            and self.length_remaining > 0
+        ):
+            while self.length_remaining > 0 and self._decoded_bytes_in_buffer() < amt:
+                data = self._raw_read(amt)
+                decoded_data = self._decode(data, decode_content, flush_decoder)
+                self._decoded_bytes.extend(decoded_data)
```

Comment:
https://docs.python.org/3/library/io.html#io.BufferedIOBase clearly says that return `b''` indicates EOF and that there is no more data to read. The point of this pull request is to returns bytes when there are bytes to read and allow using `io.BufferedIOBase` freely. So I don't understand why we only issue raw_read calls when Content-Length is set. How about this instead?

```suggestion
        if amt is not None:                                                                                                                                    
            while self._decoded_bytes_in_buffer() < amt and data:              
                data = self._raw_read(amt)                                                                                                                     
                decoded_data = self._decode(data, decode_content, flush_decoder)                                                                               
                self._decoded_bytes.extend(decoded_data)
```

Reply:
Right, this makes sense, I think it was just me overcomplicating because I didn't see the simpler solution.